src/ripple/app/tx/impl/TxQImpl.cpp

```diff
@@ -0,0 +1,969 @@
+        , minimumTxnCount_(standAlone ? 1000 : 5)
+        , txnsExpected_(minimumTxnCount_)
+        , minimumMultiplier_(500)
+        , escalationMultiplier_(minimumMultiplier_)
+        , j_(j)
+    {
+    }
+
+    /**
+    Updates fee metrics based on the transactions in the ReadView
+    for use in fee escalation calculations.
+
+    @param view View of the LCL that was just closed or received.
+    @param timeLeap Indicates that rippled is under load so fees
+    should grow faster.
+    */
+    std::size_t
+    updateFeeMetrics(Application& app,
+        ReadView const& view, bool timeLeap);
+
+    /** Used by tests only.
+    */
+    std::size_t
+    setMinimumTx(int m)
+    {
+        std::lock_guard <std::mutex> sl(lock_);
+
+        auto const old = minimumTxnCount_;
+        minimumTxnCount_ = m;
+        txnsExpected_ = m;
+        return old;
+    }
+
+    std::size_t
+    getTxnsExpected() const
+    {
+        std::lock_guard <std::mutex> sl(lock_);
+
+        return txnsExpected_;
+    }
+
+    std::uint32_t
+    getEscalationMultiplier() const
+    {
+        std::lock_guard <std::mutex> sl(lock_);
+
+        return escalationMultiplier_;
+    }
+
+    std::uint64_t
+    scaleFeeLevel(OpenView const& view) const;
+};
+
+CandidateTxn::CandidateTxn(
+    std::shared_ptr<STTx const> const& txn,
+        TxID const& txID, std::uint64_t const& feeLevel,
+            ApplyFlags const flags,
+                PreflightResult const& pfresult)
+    : txn_(txn)
+    , feeLevel_(feeLevel)
+    , txID_(txID)
+    , account_(txn->getAccountID(sfAccount))
+    , sequence_(txn->getSequence())
+    , flags_(flags)
+    , pfresult_(pfresult)
+{
+    if (txn->isFieldPresent(sfLastLedgerSequence))
+        lastValid_ = txn->getFieldU32(sfLastLedgerSequence);
+
+    if (txn->isFieldPresent(sfAccountTxnID))
+        priorTxID_ = txn->getFieldH256(sfAccountTxnID);
+}
+
+TxQAccount::TxQAccount(std::shared_ptr<STTx const> const& txn)
+    :TxQAccount(txn->getAccountID(sfAccount))
+{
+}
+
+TxQAccount::TxQAccount(const AccountID& account)
+    : account_(account)
+    , totalFees_(0)
+{
+}
+
+CandidateTxn&
+TxQAccount::addCandidate(CandidateTxn&& txn)
+{
+    auto fee = txn.getFeeLevel();
+    totalFees_ += fee;
+    auto sequence = txn.getSequence();
+
+    auto result = transactions_.emplace(std::make_pair(sequence, std::move(txn)));
+    assert(result.second);
+    assert(&result.first->second != &txn);
+
+    return result.first->second;
+}
+
+bool
+TxQAccount::removeCandidate(TxSeq const& sequence)
+{
+    return transactions_.erase(sequence) != 0;
+}
+
+CandidateTxn const*
+TxQAccount::findCandidateAt(TxSeq const& sequence) const
+{
+    auto iter = transactions_.find(sequence);
+    return iter == transactions_.end() ?
+        nullptr : &iter->second;
+}
+
+std::size_t
+FeeMetrics::updateFeeMetrics(Application& app,
+    ReadView const& view, bool timeLeap)
+{
+    std::vector<uint64_t> feeLevels;
+    for (auto const& tx : view.txs)
```

Comment:
My understanding of `ReadView::txs_type` doesn't expose a `size()` function. I'm pretty sure that `std::distance` would be even more expensive than the memory allocation due to the overhead of iterating `ReadView::txs`. (@vinniefalco, am I correct here?)

What I _could_ do is `feeLevels.reserve(txnsExpected_)`. Once the network generally accepts escalation, that value is going to remain relatively steady - or at least steady enough that we'll probably see a benefit from the hint.